src/Controller/CRUDController.php

```diff
@@ -439,12 +442,15 @@ public function batchAction(Request $request): Response
             throw new \RuntimeException('The action is not defined');
         }
 
-        $batchActions = $this->admin->getBatchActions();
-        if (!\array_key_exists($action, $batchActions)) {
-            throw new \RuntimeException(sprintf('The `%s` batch action is not defined', $action));
+        $camelizedAction = InflectorFactory::create()->build()->classify($action);
+
+        if (!$this->canBatchActionBeExecuted($action)) {
+            $finalAction = sprintf('batchAction%s', $camelizedAction);
+            throw new \RuntimeException(sprintf('A `%s::%s` method must be callable or create add a `controller` configuration', static::class, $finalAction));
         }
 
-        $camelizedAction = InflectorFactory::create()->build()->classify($action);
+        $batchAction = $this->admin->getBatchActions()[$action];
+
         $isRelevantAction = sprintf('batchAction%sIsRelevant', $camelizedAction);
 
         if (method_exists($this, $isRelevantAction)) {
```

Comment:
You forgot to support the `isRelevant` check for your "custom controller".

Reply:
I have a few bundle of sonata installed and I didn't see any place it have been implemented.

Is it use somewhere ? I think it's complicating something that could easily be done directly in the controller direclty.


```PHP
class CustomController extends CRUDController
{
  public function myCustomAction($query, $request)
  {
      // logic to test if action is relevent
      if(!$this->isRelevent($query)) {
            $this->addFlash(
                'sonata_flash_info',
                $this->trans($nonRelevantMessage, [], 'SonataAdminBundle')
            );
            
            return $this->redirectToList();
      }
      
      // continue logic
  }
}
```

I have extended the CRUDController just to be consistent with the method call, I think we should extract some method method from the CRUD controller and put them in a trait that could be use for both CRUDController and custom action. It's already an issue with some of the function that could be useful in Admin related action (like renderWithExtraParams, isXmlHttpRequest, handleModelManagerThrowable, etc.)

Reply:
Dunno, never used it too. I wonder if we should deprecate (in another pr) this isRelevant feature.

Reply:
the $this->isRevelant() dooes not exists as is, it was to show how to put it in the controller instead of a hook.

But yes we should probably deprecate the hook call so dev move it to their controller. Less logic to maintain on sonata side.